docs.google.com: Can't type with Devanagari input source

rniwa opened the pull request:
#### 1322eea39d8f8bcaa4d562c9edca2f5016c7b6bc
<pre>
docs.google.com: Can&apos;t type with Devanagari input source
<a href="https://bugs.webkit.org/show_bug.cgi?id=295763">https://bugs.webkit.org/show_bug.cgi?id=295763</a>

Reviewed by Wenson Hsieh.

The bug was caused by WebKit dispatching keydown event after composition and input events when
handling key events with input methods.

This PR fixes the ordering of these events by deferring processing of setMarkedText until after
the corresponding keydown event is fired in WebContent process. In order to maintain the correct
ordering of input method processing of key events, we also defer calls to interpretKeyEvent
while another input method processing is taking place.

Because this behavior change is rather risky, we hide the new behavior behind a new runtime flag
and only enable it on docs.google.com as a quirk for now. We determine whether the quirk is needed
or not when computing EditorState and sends a boolean back to UI process to enable the behavior.
We expect to eventually ship the new behavior everywhere.

Unfortunately no new tests for now since we don&apos;t have an infrastructure to test this code yet.
Will follow up.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Added a new runtime flag.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::internalKeyEvent): When we&apos;re in the middle of processing a composition,
explicitly call Editor::handleKeyboardEvent to process keypress commands in the event. Without
this change, we would drop deferred composition updates, etc... on floor.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::inputMethodUsesCorrectKeyEventOrder const): Added.
(WebCore::handleGoogleQuirks): Added. Sets inputMethodUsesCorrectKeyEventOrder to true when the
top level document&apos;s host is docs.google.com.

* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
(WebCore::QuirksData):

* Source/WebCore/platform/KeypressCommand.h:
(WebCore::KeypressCommand::KeypressCommand): Added a new constructor for setMarkedText. To defer
setMarkedText, store composition underline &amp; highlight, selected range, and replacement range.
This constructor is also now used by the IPC message decoder for KeypressCommand.

* Source/WebKit/Shared/EditingRange.h:
(WebKit::EditingRange::EditingRange): Added a new constructor which takes WebCore::CharacterRange.
(WebKit::EditingRange::toCharacterRange const): Added.

* Source/WebKit/Shared/EditorState.h:
(WebKit::EditorState):
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getSelectedRangeAsync): Now returns selected range and composition range.

* Source/WebKit/UIProcess/WebPageProxy.h:

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
(WebKit::WebViewImpl): Updated m_collectedKeypressCommands to be an optional of a Vector of
KeypressCommand instead of a CheckedPtr to a struct which contains the Vector. Also added
m_stagedMarkedRange which is a selection set by currently deferred setMarkedText, and
m_interpretKeyEventHoldingTank which is a Vector of completion handlers which will execute
deferred calls to WebViewImpl::interpretKeyEvent.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::collectKeyboardLayoutCommandsForEvent):
(WebKit::WebViewImpl::interpretKeyEvent): Defer all other calls to interpretKeyEvent and
setMarkedText while we&apos;re processing the current event via handleEventByInputMethod. Execute
the deferred setMarkedText via KeypressCommand and call all the completion handlers in
m_interpretKeyEventHoldingTank once the current call to interpretKeyEvent finishes.

(WebKit::WebViewImpl::doCommandBySelector):
(WebKit::WebViewImpl::insertText):
(WebKit::WebViewImpl::selectedRangeWithCompletionHandler): When m_stagedMarkedRange is set,
return the NSRange which reflects the effect of m_stagedMarkedRange.

(WebKit::WebViewImpl::setMarkedText): Defer this call to setMarkedText if we&apos;re in the middle
of processing key event via WebViewImpl::interpretKeyEvent. setMarkedText will be executed
later via KeypressCommand by WebPage::executeKeypressCommandsInternal.

(WebKit::WebViewImpl::keyUp): Remove the assertion that if handledByInputMethod is true then
commands is empty since this is no longer true after this PR.
(WebKit::WebViewImpl::keyDown): Ditto.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::WebEditorClient::didDispatchInputMethodKeydown): Added.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getSelectedRangeAsync): Now sends back the composition range in addition to
the selected range.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::getPlatformEditorState const): Decide whether new behavior should be enabled
or not for this editable area.
(WebKit::WebPage::executeKeypressCommandsInternal): Added the processing of setMarkedText:.

Canonical link: <a href="https://commits.webkit.org/297270@main">https://commits.webkit.org/297270@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7670bcbf028cb7cc086ef7e519017999253bf95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111133 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117164 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61401 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113095 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31480 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39381 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/84492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114080 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25149 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100052 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64938 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24494 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18196 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60984 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103624 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/94529 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18260 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120134 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109686 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38182 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28378 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93431 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38558 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96327 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93255 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38337 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16075 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34167 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17920 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38071 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43548 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133962 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37736 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36136 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41069 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39438 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/QuirksData.h:
```diff
@@ -28,6 +28,7 @@
 namespace WebCore {
 
 struct WEBCORE_EXPORT QuirksData {
+    bool inputMethodUsesCorrectKeyEventOrder : 1 { false };
```
Nit - it looks like this first section is for `is(Domain/Website)` flags. Maybe this could be moved to just before `bool hasBrokenEncryptedMediaAPISupportQuirk`?

rniwa replied:
Moved it right beneath `inputMethodUsesCorrectKeyEventOrder` below.

whsieh commented on Source/WebCore/platform/KeypressCommand.h:
```diff
@@ -26,6 +26,9 @@
 #ifndef KeypressCommand_h
 #define KeypressCommand_h
 
+#include "CharacterRange.h"
+#include "CompositionHighlight.h"
+#include "CompositionUnderline.h"
 #include <wtf/Assertions.h>
```
Hm... these 3 headers are in `WebCore/editing`, but this header (`KeypressCommand.h`) is in `WebCore/platform`.

It doesn't seem unreasonable to move those into `platform/`, maybe with more generic names (or we could go the other way and try to move `KeypressCommand` out of `platform/`, though it seems a bit tricky to disentangle with `PlatformKeyboardEvent`).

I think we can do that separately though.

rniwa replied:
Sure, will follow up.

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5127,21 +5126,67 @@ static bool eventKeyCodeIsZeroOrNumLockOrFn(NSEvent *event)
+#if PLATFORM(MAC)
+    if (m_page->editorState().inputMethodUsesCorrectKeyEventOrder) {
+        if (m_collectedKeypressCommands) {
+            m_interpretKeyEventHoldingTank.append([weakThis = WeakPtr { *this }, capturedEvent = retainPtr(event), capturedBlock = makeBlockPtr(completionHandler)]() {
```
```suggestion
            m_interpretKeyEventHoldingTank.append([weakThis = WeakPtr { *this }, capturedEvent = retainPtr(event), capturedBlock = makeBlockPtr(completionHandler)] {
```

whsieh commented on Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
```diff
@@ -177,6 +178,7 @@
     getPlatformEditorStateCommon(frame, result);
 
     result.canEnableAutomaticSpellingCorrection = result.isContentEditable && frame.protectedEditor()->canEnableAutomaticSpellingCorrection();
+    result.inputMethodUsesCorrectKeyEventOrder = frame.settings().inputMethodUsesCorrectKeyEventOrder() || (frame.document() && frame.document()->quirks().inputMethodUsesCorrectKeyEventOrder());
```
```suggestion
    result.inputMethodUsesCorrectKeyEventOrder = frame.settings().inputMethodUsesCorrectKeyEventOrder() || (frame.document() && frame.protectedDocument()->quirks().inputMethodUsesCorrectKeyEventOrder());
```

rniwa replied:
I'll add a RefPtr local variable.

rniwa committed:
docs.google.com: Can't type with Devanagari input source
https://bugs.webkit.org/show_bug.cgi?id=295763

Reviewed by Wenson Hsieh.

The bug was caused by WebKit dispatching keydown event after composition and input events when
handling key events with input methods.

This PR fixes the ordering of these events by deferring processing of setMarkedText until after
the corresponding keydown event is fired in WebContent process. In order to maintain the correct
ordering of input method processing of key events, we also defer calls to interpretKeyEvent
while another input method processing is taking place.

Because this behavior change is rather risky, we hide the new behavior behind a new runtime flag
and only enable it on docs.google.com as a quirk for now. We determine whether the quirk is needed
or not when computing EditorState and sends a boolean back to UI process to enable the behavior.
We expect to eventually ship the new behavior everywhere.

Unfortunately no new tests for now since we don't have an infrastructure to test this code yet.
Will follow up.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Added a new runtime flag.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::internalKeyEvent): When we're in the middle of processing a composition,
explicitly call Editor::handleKeyboardEvent to process keypress commands in the event. Without
this change, we would drop deferred composition updates, etc... on floor.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::inputMethodUsesCorrectKeyEventOrder const): Added.
(WebCore::handleGoogleQuirks): Added. Sets inputMethodUsesCorrectKeyEventOrder to true when the
top level document's host is docs.google.com.

* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
(WebCore::QuirksData):

* Source/WebCore/platform/KeypressCommand.h:
(WebCore::KeypressCommand::KeypressCommand): Added a new constructor for setMarkedText. To defer
setMarkedText, store composition underline & highlight, selected range, and replacement range.
This constructor is also now used by the IPC message decoder for KeypressCommand.

* Source/WebKit/Shared/EditingRange.h:
(WebKit::EditingRange::EditingRange): Added a new constructor which takes WebCore::CharacterRange.
(WebKit::EditingRange::toCharacterRange const): Added.

* Source/WebKit/Shared/EditorState.h:
(WebKit::EditorState):
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getSelectedRangeAsync): Now returns selected range and composition range.

* Source/WebKit/UIProcess/WebPageProxy.h:

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
(WebKit::WebViewImpl): Updated m_collectedKeypressCommands to be an optional of a Vector of
KeypressCommand instead of a CheckedPtr to a struct which contains the Vector. Also added
m_stagedMarkedRange which is a selection set by currently deferred setMarkedText, and
m_interpretKeyEventHoldingTank which is a Vector of completion handlers which will execute
deferred calls to WebViewImpl::interpretKeyEvent.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::collectKeyboardLayoutCommandsForEvent):
(WebKit::WebViewImpl::interpretKeyEvent): Defer all other calls to interpretKeyEvent and
setMarkedText while we're processing the current event via handleEventByInputMethod. Execute
the deferred setMarkedText via KeypressCommand and call all the completion handlers in
m_interpretKeyEventHoldingTank once the current call to interpretKeyEvent finishes.

(WebKit::WebViewImpl::doCommandBySelector):
(WebKit::WebViewImpl::insertText):
(WebKit::WebViewImpl::selectedRangeWithCompletionHandler): When m_stagedMarkedRange is set,
return the NSRange which reflects the effect of m_stagedMarkedRange.

(WebKit::WebViewImpl::setMarkedText): Defer this call to setMarkedText if we're in the middle
of processing key event via WebViewImpl::interpretKeyEvent. setMarkedText will be executed
later via KeypressCommand by WebPage::executeKeypressCommandsInternal.

(WebKit::WebViewImpl::keyUp): Remove the assertion that if handledByInputMethod is true then
commands is empty since this is no longer true after this PR.
(WebKit::WebViewImpl::keyDown): Ditto.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::WebEditorClient::didDispatchInputMethodKeydown): Added.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getSelectedRangeAsync): Now sends back the composition range in addition to
the selected range.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::getPlatformEditorState const): Decide whether new behavior should be enabled
or not for this editable area.
(WebKit::WebPage::executeKeypressCommandsInternal): Added the processing of setMarkedText:.

Canonical link: https://commits.webkit.org/297270@main
